src/ilasm/CMakeLists.txt

```diff
@@ -52,7 +52,9 @@ if(CLR_CMAKE_PLATFORM_UNIX)
   # Clang also produces a bad-codegen on this prebuilt file with optimization.
   # https://github.com/dotnet/coreclr/issues/2305
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-delete-non-virtual-dtor")
-  add_compile_options(-Wno-deprecated-register)
+  if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+    add_compile_options(-Wno-deprecated-register)
```

Comment:
Has it made conditional because this warning is not supported in gcc, or not required by gcc? If it is not supported, then maybe `add_compile_options(-Wno-deprecated-declarations)` is the correct replacement?

This warning comes from `asmparse.cpp` (renamed from `asmparse.c`) that uses `register` keyword. The `register` keyword is deprecated in C++11 and removed in C++17. The keyword is still reserved (for warning purposes?). I think certain version of GCC will also complain about it.

Perhaps at later point, `asmparse.cpp` could be changed back to `.c` with  the accompanying `grammar_before/after.cpp`, which could be header files instead of `.cpp` in first place? There are other examples of `#include <path/file.cpp>` which can be found by `git grep -l "#include.*\.cpp"`.

Reply:
I have not seen any warnings from gcc. This warning is not supported by gcc. That waa the reason for the change.

Reply:
TL;DR version: could you unconditionally try to use `-Wno-register`?

```diff
-   add_compile_options(-Wno-deprecated-register)
+   add_compile_options(-Wno-register)
```

#### C++

Did some testing with `std=c++17` with recent version of g++:

```sh
[am11@vdnframe3 ~]$ g++ --version
g++ (FreeBSD Ports Collection) 8.2.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

With `CXX=g++`,

```sh
$CXX -x c++ -std=c++17 -O -Wall -o myapp.exe - << EOF
#include <iostream>
int main() {
register char* hwString = (char*)"Hello World\n";
std::cout << hwString;
return 0;
}
EOF
```
We get:

> <stdin>: In function 'int main()':
> <stdin>:3:16: warning: ISO C++17 does not allow 'register' storage class specifier [-Wregister]

With `CXX=clang++` (version `FreeBSD clang version 6.0.0 (tags/RELEASE_600/final 326565) (based on LLVM 6.0.0)`), we get:

> <stdin>:3:1: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
> register char* t = (char*)"Hello World";
> ^~~~~~~~~
> 1 error generated.

If we use `-Wno-register` as in `$CXX -x c++ -std=c++17 -O -Wall -Wno-register -o myapp.exe`, it makes both compilers happy.


#### C

With `CC=gcc` or `CC=clang`, and `-std=c99`, we get no warnings from either compilers; both are fine with the usage of `register` keyword:

```sh
[am11@vdnframe3 ~]$ $CC -x c -std=c99 -O -Wall -o myapp.exe - << EOF
#include <stdio.h>
int main() {
register char* hwString = (char*)"Hello World\n";
printf("%s", hwString);
return 0;
}
EOF
```

Reply:
done